### PR TITLE
fix(type): add bool and List[bool] for join's on input

### PR DESCRIPTION
**Pull request by hongbo-miao**
### What changes were proposed in this pull request?

I think `join`'s `on` input can be `bool` or `List[bool]` type. For example, the demo in the comment is a valid demo:

https://github.com/apache/spark/blob/309638eeefbfb13dae8dbded0279bf44390389ee/python/pyspark/sql/dataframe.py#L2107-L2108

### Why are the changes needed?

I originally got this typing error in my IDE:

image

The command joins two table successfully on different columns, however, the typing definition is wrong I think.

### Does this PR introduce _any_ user-facing change?

Yes, I am using `pyspark==3.3.0`.

### How was this patch tested?

After adding `bool` and `List[bool]`, the typing error is gone.

### CODE REVIEW

File: python/pyspark/sql/dataframe.py [hunk 1]
```diff
@@ -2044,7 +2044,7 @@ def crossJoin(self, other: "DataFrame") -> "DataFrame":
     def join(
         self,
         other: "DataFrame",
-        on: Optional[Union[str, List[str], Column, List[Column]]] = None,
+        on: Optional[Union[str, List[str], bool, List[bool], Column, List[Column]]] = None,
```
Comment (HyukjinKwon) [inline python/pyspark/sql/dataframe.py:2047]: That's actually an error from IDE (which assumes that the built-in functions always return `bool`). The expected types here are correct in fact.

Comment (HyukjinKwon) [reply python/pyspark/sql/dataframe.py:2047]: Yes.

Comment (Hongbo-Miao) [reply python/pyspark/sql/dataframe.py:2047]: I see. I found something interesting.

Intellij IDEA has no issue with pandas `join`'s `on`.
image
But seems only has issue for pyspark.

Here is the way pandas implement:

- https://github.com/pandas-dev/pandas/blob/main/pandas/core/frame.py#L9837-L9840
- https://github.com/pandas-dev/pandas/blob/main/pandas/_typing.py#L121

Comment (HyukjinKwon) [reply python/pyspark/sql/dataframe.py:2047]: Okay, that's interesting. cc @zero323 fyi.

Comment (zero323) [reply python/pyspark/sql/dataframe.py:2047]: @HyukjinKwon  As far as I can tell, there is no real difference here. The difference between Pandas and Spark check, is most likely related to the missing stubs for the former one. If I use environment without `pandas-stubs` things type check in PyCharm 

without pandas stubs

If I choose one with `pandas-stubs` installed, I get 

with pandas stubs

which is the same category of failure as for PySpark code.

Given mypy as a reference, this is an expected false positive (see https://github.com/python/mypy/issues/2783).

On a side note ‒ Pandas and PySpark joins shown on the screenshots are not even remotely equivalent.

**Pull request closed without merging**